### ISSUE TITLE
api-core bundle usage

### ISSUE THREAD

**Issue by xleliberty** [sulu/sulu#5151]
Hi there, 
i was wondering if anybody have already tested adding api-core bundle to replace fos rest bundle implementation ? 
It would be a great addition to ease headless usage ( graphQL and more ) 

Thanks 
Xavier

**Comment by nnatter** [comment]
Hey, thanks for your issue! I guess you are talking about the `api-platform/core` package?

Personally, I do not have much experience with the API platform. Judging from the `Getting started` guide only, the whole thing feels quite comprehensive and opinionated. So I am not sure if it is possible to smoothly integrate this into the existing codebase. But I may be mistaken here because of missing experience 🙂 

How would you expect such an integration to work? What entities would you expect to be accessible through the API platform? 
Do you have experience with using the API platform in big projects? Can we use only certain parts without causing unexpected problems?

**Comment by xleliberty** [comment]
Hi, 
api-platform/core is now the "defacto" standard for api projects in symfony so it would be a replacement for fos-rest-bundle.
we use it in a quite big project since first versions and it's really versatile and easy to use. 

it would be possible to think about a new version of Sulu Api with it while keeping fos-rest for compat.

But it seems that nobody have tried it on sulu projects yet ? 

Regards
Xavier

**Comment by Warxcell** [comment]
Yeah, that looks like good plan. Next major version - Introduce API Platform with deprecating FOS Rest.
And Next Major - Remove FOS Rest.

**Comment by danrot** [comment]
Unfortunately the FOSRestBundle is taking a turn we also don't like (in particular I am talking about [this issue](https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1954)), but we have decided to stay with it for now, because the BC break would be too big.

I don't know what we will do about that in the future, but API Platform is definitely something we have to have a closer look at. But we have already encountered a few things that feel strange, e.g. that [the value of `item` seems to be hardcoded for HAL](https://github.com/api-platform/core/blob/97fa954/src/Hal/Serializer/CollectionNormalizer.php#L84). This contradicts with our current implementation (we use values like `contacts`, `accounts` and so on). While I think they might be open to change that, we also heard from different people that API platform feels a bit like rapid prototyping, which would be a counterargument from our perspective, since we have to keep that code for quite some time.

We have also already thought about implementing the REST stuff very specialized on our API on our own, so that we don't have another dependency which might make problems when updating stuff. In that case you could also add whatever you like, including API platform, if the do not hardcode the `item` value (or we change our implementation because of that).

**Comment by Warxcell** [comment]
@danrot You are looking specifically at HAL format. there are multiple formats. JSON-LD, JSON, HAL and etc. You can add even your own format/content-type.

You can also use a DTOs just for API which are separated from Entities. That way it won't feel like Rapid Prototyping.
You can even implement your own Data Providers and get data from wherever you want.

**Comment by danrot** [comment]
> @danrot You are looking specifically at HAL format. there are multiple formats. JSON-LD, JSON, HAL and etc. You can add even your own format/content-type.

Yes, because that's exactly what we use, therefore it is currently the only important one to me :wink: 

> You can also use a DTOs just for API which are separated from Entities. That way it won't feel like Rapid Prototyping.
> You can even implement your own Data Providers and get data from wherever you want.

That already sounds so complicated, that I am wondering if just implementing a custom route loader which registers routes for `GET`, `POST` and so on and using the Symfony Serializer would not cover almost everything we need without having more code dependencies that are out of our control.

But as said, must have a closer look, that's just an initial feeling.

> Also two of maintainers of Api-Platform is also part of Symfony team itself.

Interestingly API Platform is still not part of Symfony itself :grimacing:

**Comment by Warxcell** [comment]
As the team wishes. But I would compare implementing own API solution to be same as implementing own ORM. 
In the end there will be always something missing or it will be too hard to maintain.

**Comment by danrot** [comment]
Depends on the needs. Also the FOSRestBundle comes with a lot of functionality that we never used, but we still had issues on every major Symfony upgrade, because FOSRestBundle (but also other dependencies) have not updated yet. It's as always a tradeoff, and I want to make sure that we get enough out of relying on this dependency. I would be fairly confident to say the REST functionality what we are using is not too much work to implement on our own, and then we would not have to check all our dependencies on every update (which is also a thing hard to maintain). That's why I am not so sure about this at the moment.

**Comment by chirimoya** [comment]
As already mentioned, for 3.0 we will certainly look at the topic in detail again and all options are on the table.

**Comment by xleliberty** [comment]
ok, because to me it would be possible to keep fosrest for "legacy" api and rely on api-core to implement the next one ;)

regards
Xavier

**Comment by jorismak** [comment]
Is this about 'allowing' Sulu users to install the `api` recipe, and use Api Platform to create their own API parts, completely separate from Sulu?

Or is it to replace certain Sulu internals with api-platform instead of FOSRestBundle?

We use api-platform internally here a lot, and 'simply adding it' to a Sulu project would be awesome. Specially if we can then use it to generate simple REST endpoints for our custom entities, that the Sulu-admin can then use.
But I didn't try this yet, so I don't know if it's already possible, or if there are conflicts around.

**Comment by alexander-schranz** [comment]
@jorismak we had many discussion about this topic in the past. We will not use Api Platform in Sulu core because we want to keep the dependency tree small as possible. Also the case why we planning to remove FOSRestBundle from the core as it just make it harder to keep things uptodate with latest Symfony versions.

Still we want make sure developers using Sulu can use whatever they want to use to create der API. This means whatever they want to stay pure Symfony, FosRest, Api Platform or whatever they want. At current state Sulu uses the json hal similar format which is also supported by Api Platform. But Api Platform at the state 2020 didn't support naming the attributes. This might have changed the last year but I didn't yet have a relook at it.

So to be clear Sulu Admin only requires that you return the data in the expected format: https://github.com/api-platform/core/issues/4141. What tools you use to get it into that format is up to you.

**Comment by jorismak** [comment]
I was hoping for the 'do not get into the way' idea 😊.
If it's not in the way, then I don't really care what sulu uses internally
to be honest.

And these days, with modern symfony attributes, getting your body
deserialized, validated and injected into a controller is a one liner. The
same for returning an object and seializing it into a format or your
choice.
For quick and easy json stuff, no bundle is required. Documenting it is
something else though 😉.
